source/slang/slang-lower-to-ir.cpp

```diff
@@ -6429,10 +6434,35 @@ struct StmtLoweringVisitor : StmtVisitor<StmtLoweringVisitor>
 
     void visitBlockStmt(BlockStmt* stmt)
     {
-        // To lower a block (scope) statement,
-        // just lower its body. The IR doesn't
-        // need to reflect the scoping of the AST.
+        // Blocks affect the `defer` statement, so we need to track their scopes
+        // with a decorator in the IR.
+        bool prevScopeActive = context->scopeEndBlockActive;
+        IRBlock* prevScopeEndBlock = context->scopeEndBlock;
+
+        auto builder = getBuilder();
+        context->scopeEndBlockActive = false;
+        context->scopeEndBlock = builder->createBlock();
+
+        // To lower a block (scope) statement, just lower its body.
         lowerStmt(context, stmt->body);
+
+        if (context->scopeEndBlockActive)
```

Comment:
you should be able to call `context->scopeEndBlock->hasUses()` instead of maintaining a separate `scopeEndBlockActive` variable.